server/src/main/java/io/crate/replication/logical/action/ShardChangesAction.java

```diff
@@ -231,6 +237,34 @@ protected ShardsIterator shards(ClusterState state, Request request) {
                 request.shardId().id()
             ).activeInitializingShardsRandomIt();
         }
+
+        /**
+         * A publisher should not send updates when a subscriber polls changes of an index
+         * where not all shards are active yet, restore will fail if primaries are not (yet) assigned.
+         * <p>
+         * We include such indices to the PublicationsStateAction.Response as otherwise subscriber
+         * can treat "not ready to replicate" index as deleted and turn a table into a regular one on its side.
+         * For example, this can happen on a publisher cluster restart.
+         *
+         * @return boolean flag indicating that listener is failed and async callee can return right away.
+         */
+        private boolean ensureAllPrimariesActive(ClusterState state,
+                                                 ShardId shardId,
+                                                 @Nullable ActionListener<Response> listener) {
+            var routingTable = state.routingTable().index(shardId.getIndexName());
+            assert routingTable != null : "routingTable must not be null";
+            if (routingTable.allPrimaryShardsActive() == false) {
```

Comment:
It's WIP because some tests are failing with an assertion error. If I add NPE check, test pass but I want to fugure out why we can have null routing table. 

Also, cluster state is completely empty when `routingTable` is null, so maybe test setup issue